File: fn/src/main/java/org/sqlbuilder/fn/objects/SemType.java
package org.sqlbuilder.fn.objects;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.common.Utils;
import org.sqlbuilder.fn.HasID;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import edu.berkeley.icsi.framenet.SemTypeType;

public class SemType implements HasID, Insertable
{
	public static final Set<SemType> SET = new HashSet<>();

	private final int semtypeid;

	private final String name;

	private final String abbrev;

	private final String definition;

	public static SemType make(final SemTypeType type)
	{
		var t = new SemType(type);
		SET.add(t);
		return t;
	}

	private SemType(final SemTypeType type)
	{
		this.semtypeid = type.getID();
		this.name = type.getName();
		this.abbrev = type.getAbbrev();
		this.definition = type.getDefinition();
	}

	// A C C E S S

	public int getID()
	{
		return semtypeid;
	}

	public String getName()
	{
		return name;
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		SemType that = (SemType) o;
		return semtypeid == that.semtypeid;
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(semtypeid);
	}

	// O R D E R

	public static final Comparator<SemType> COMPARATOR = Comparator.comparing(SemType::getName).thenComparing(SemType::getID);

	// I N S E R T

	@Override
	public String dataRow()
	{
		return String.format("%d,'%s','%s','%s'", //
				semtypeid, //
				name, //
				abbrev, //
				Utils.escape(definition));
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[SEMTYPE semtypeid=%s name=%s]", this.semtypeid, this.name);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/objects/SemType.kt
package org.sqlbuilder.fn.objects

import edu.berkeley.icsi.framenet.SemTypeType
import org.sqlbuilder.common.HasID
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.Utils.escape
import java.util.*

class SemType private constructor(
    type: SemTypeType,
) : HasID, Insertable {

    val iD: Int = type.getID()

    val name: String = type.getName()

    private val abbrev: String = type.getAbbrev()

    private val definition: String = type.getDefinition()

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as SemType
        return iD == that.iD
    }

    override fun hashCode(): Int {
        return Objects.hash(iD)
    }

    // I N S E R T

    override fun dataRow(): String {
        return "$iD,'$name','$abbrev','${escape(definition)}'"
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[SEMTYPE semtypeid=$iD name=$name]"
    }

    companion object {

        @JvmField
        val COMPARATOR: Comparator<SemType> = Comparator
            .comparing<SemType, String> { it.name }
            .thenComparing<Int> { it.iD }

        @JvmField
        val SET = HashSet<SemType>()

        @JvmStatic
        fun make(type: SemTypeType): SemType {
            val t = SemType(type)
            SET.add(t)
            return t
        }
    }
}
